src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPEditorImprovementsEnterBetweenBracesHandler.java

```diff
@@ -0,0 +1,200 @@
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Ref;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.codeStyle.CodeStyleSettings;
+import com.redhat.devtools.lsp4ij.LSPIJEditorUtils;
+import com.redhat.devtools.lsp4ij.client.features.LSPEditorBehaviorFeature;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * This provides a workaround for <a href="https://youtrack.jetbrains.com/issue/IJPL-159454">IJPL-159454</a>.
+ */
+public class LSPEditorImprovementsEnterBetweenBracesHandler extends EnterBetweenBracesHandler {
+    private boolean enterAfterOpenBrace = false;
+    private Character openBraceCharacter = null;
+    private boolean enterBeforeCloseBrace = false;
+    private Character closeBraceCharacter = null;
+
+    @Override
+    public Result preprocessEnter(
+            @NotNull PsiFile file,
+            @NotNull Editor editor,
+            @NotNull Ref<Integer> caretOffset,
+            @NotNull Ref<Integer> caretAdvance,
+            @NotNull DataContext dataContext,
+            @Nullable EditorActionHandler originalHandler) {
+        if (LSPIJEditorUtils.isSupportedTextMateFile(file) &&
+            CodeInsightSettings.getInstance().SMART_INDENT_ON_ENTER &&
+            LSPEditorBehaviorFeature.enableTextMateEnterBetweenBracesFix(file)) {
+            CaretModel caretModel = editor.getCaretModel();
+            int offset = caretModel.getOffset();
+            if (offset > 0) {
+                Document document = editor.getDocument();
+                CharSequence charsSequence = document.getCharsSequence();
+                char previousCharacter = charsSequence.charAt(offset - 1);
+                if (LSPIJEditorUtils.isOpenBraceCharacter(file, previousCharacter)) {
+                    enterAfterOpenBrace = true;
+                    openBraceCharacter = previousCharacter;
+                    closeBraceCharacter = LSPIJEditorUtils.getCloseBraceCharacter(file, openBraceCharacter);
+                    if (closeBraceCharacter != null) {
+                        char nextCharacter = charsSequence.charAt(offset);
+                        if (closeBraceCharacter == nextCharacter) {
+                            enterBeforeCloseBrace = true;
+                        }
+                    }
+                }
+            }
+        }
+
+        return super.preprocessEnter(file, editor, caretOffset, caretAdvance, dataContext, originalHandler);
+    }
+
+    @Override
+    public Result postProcessEnter(
+            @NotNull PsiFile file,
+            @NotNull Editor editor,
+            @NotNull DataContext dataContext) {
+        if (LSPIJEditorUtils.isSupportedTextMateFile(file) &&
+            CodeInsightSettings.getInstance().SMART_INDENT_ON_ENTER &&
+            LSPEditorBehaviorFeature.enableTextMateEnterBetweenBracesFix(file)) {
```

Comment:
Immediate feature flag check so if this feature is disabled in client config, this is a no-op.